### PR TITLE
fix/81553 Add warnings for illegal file names in role

### LINKED ISSUE #81553: ansible-galaxy install of roles with Java inner classes fails due to $ in the file name
### Summary

When I try to install a role that contains files with `$` in their names (such as Java class names), it fails because we don't allow `$` in the file names because file names could get evaluated via `os.path.expandvars`.

See https://github.com/ansible/galaxy/issues/271 for the original context

The error message is "Not a directory" (errno 20).

### Issue Type

Bug Report

### Component Name

role

### Ansible Version

```console
$ ansible --version
ansible [core 2.14.5]
  config file = /Users/atsalolikhin/.ansible.cfg
  configured module search path = ['/Users/atsalolikhin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/atsalolikhin/py-3.9.13/lib/python3.9/site-packages/ansible
  ansible collection location = /Users/atsalolikhin/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/atsalolikhin/py-3.9.13/bin/ansible
  python version = 3.9.13 (main, Jul 20 2022, 17:45:35) [Clang 13.0.0 (clang-1300.0.29.3)] (/Users/atsalolikhin/py-3.9.13/bin/python3.9)
  jinja version = 3.0.3
  libyaml = True
```

### Configuration

 N/A
 

### OS / Environment

N/A

### Steps to Reproduce

Add this to your `requirements.yaml`:

```yaml
- src: https://gitlab.com/atsaloli/ansible-galaxy-issue-271.git
  version: main
  scm: git
```

and then try to install roles, with e.g., `ansible-galaxy install -r requirements.yml`

### Expected Results

I expected the role to be installed.

### Actual Results

```console
The role install failed.


$ ansible-galaxy install -r requirements.yml --force
Starting galaxy role install process
- extracting ansible-galaxy-issue-271 to /home/atsaloli/.ansible/roles/ansible-galaxy-issue-271
[WARNING]: - ansible-galaxy-issue-271 was NOT installed successfully: Could not update files in /home/atsaloli/.ansible/roles/ansible-galaxy-issue-271:
[Errno 20] Not a directory: '/home/atsaloli/.ansible/roles/ansible-galaxy-issue-271/files/udf/CMC_VWAP.class'
ERROR! - you can use --ignore-errors to skip failed roles and finish processing the list.
$
```
```


### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by atsalolikhin-spokeo**
##### SUMMARY

Does not fix #81553 (but makes it a little better).

##### ISSUE TYPE

- Bugfix Pull Request
 

##### ADDITIONAL INFORMATION

See #81553

### CODE REVIEW

**Comment by atsalolikhin-spokeo** [comment]
@bcoca Is this something like what you had in mind?  I'm just learning Python so feel free to steer me in the right direction if there is anything off about this PR.  

I didn't test my changes because the README says you install from pip and my changes are not in pip.  I wasn't sure how to install Ansible from the source tree.

However the PR passed the CI test:  78 successful checks.

Thanks!

**Comment by bcoca** [comment]
I would change those to warnings (`display.warning()`) to keep it backwards compatible, but give users at least a hint why the files were not extracted.

**Comment by atsalolikhin-spokeo** [comment]
@bcoca Thanks!  I added `display.warning()`.

I'm not sure I correctly understood your comment https://github.com/ansible/galaxy/issues/271#issuecomment-1686846546 saying this should not be an unhandled exception.

I am raising `RuntimeError` with an explanation now rather than letting Python exit with Errno 20 and a cryptic and potentially misleading "Not a directory" error message.

Is that what you had in mind here?

**Comment by bcoca** [comment]
The 'unexpected exception' message comes from our 'catchall' and is normally indication of a bug, for this one i think we should handle the issue more gracefully and return a more pertinent error message that allows the user to fix/bypass the issue.

Your current code does avoid this but it changes behavior by preventing the install of the role at all vs just skipping the 'bad files', this is why i suggested above you change it into a warning. The ultimate fix for the original error would capture the existing exception and reraise it in an `AnsibleError` with appropriate messaging.

**Comment by bcoca** [comment]
This is the loop which generates the error reported
https://github.com/ansible/ansible/pull/81555/files#diff-b99ff71e25ad80853389fe308d3e0693231cfb21e28ecaf7f80bd4a3ee23ade1L317

it should either have the same/similar code and skip 'bad members' or handle the exception and either reraise or skip (cause we know why we didnt create the dir).

File: lib/ansible/galaxy/role.py [hunk 1]
```diff
@@ -371,13 +371,20 @@ def install(self):
...
                                 for n_part in n_parts:
-                                    # TODO if the condition triggers it produces a broken installation.
-                                    # It will create the parent directory as an empty file and will
-                                    # explode if the directory contains valid files.
-                                    # Leaving this as is since the whole module needs a rewrite.
-                                    if n_part != '..' and not n_part.startswith('~') and '$' not in n_part:
-                                        n_final_parts.append(n_part)
+                                    if n_part == '..':
+                                        display.warning(f"Illegal filename '{n_part}': '..' is not allowed")
+                                        raise RuntimeError
```
Comment (bcoca) [inline lib/ansible/galaxy/role.py]: remove the raise RuntimeError and just add a continue (next item in loop)

**Review by bcoca — CHANGES_REQUESTED** [review]
this will also need a changelog and tests

**Comment by atsalolikhin-spokeo** [comment]
@bcoca Thank you very much. Could you please point me in the right direction for how to install Ansible from source code? I want to be able to test my changes end-to-end.

**Comment by bcoca** [comment]
https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html#installing-for-development
^ personally i use the clone, but any option should work

**Comment by atsalolikhin-spokeo** [comment]
# Validation

## Before

```
$ ansible-galaxy install -r requirements.yml --force
Starting galaxy role install process
- extracting ansible-galaxy-issue-271 to /home/atsaloli/.ansible/roles/ansible-galaxy-issue-271
[WARNING]: - ansible-galaxy-issue-271 was NOT installed successfully: Could not update files in /home/atsaloli/.ansible/roles/ansible-galaxy-issue-271:
[Errno 20] Not a directory: '/home/atsaloli/.ansible/roles/ansible-galaxy-issue-271/files/udf/CMC_VWAP.class'
ERROR! - you can use --ignore-errors to skip failed roles and finish processing the list.
$
```

## After

```
[/tmp] [py-3.11] $ ansible-galaxy install -r requirements.yaml
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying
the Ansible engine, or trying out features under development. This is a rapidly changing source of code and can become
unstable at any point.
Starting galaxy role install process
- extracting ansible-galaxy-issue-271 to /Users/atsalolikhin/.ansible/roles/ansible-galaxy-issue-271
[WARNING]: Illegal filename 'CMC_VWAP$CMC_VWAP_proc.class': names cannot contain '$'
[WARNING]: - ansible-galaxy-issue-271 was NOT installed successfully: Could not update files in
/Users/atsalolikhin/.ansible/roles/ansible-galaxy-issue-271: [Errno 20] Not a directory:
'/Users/atsalolikhin/.ansible/roles/ansible-galaxy-issue-271/files/udf/CMC_VWAP.class'
ERROR! - you can use --ignore-errors to skip failed roles and finish processing the list.
[/tmp] [py-3.11] $
```

The behavior is exactly the same as before, except it adds a warning to give a hint to the user.

```
[WARNING]: Illegal filename 'CMC_VWAP$CMC_VWAP_proc.class': names cannot contain '$'
```

In both cases, role installation fails halfway through.  

I feel this changeset accomplishes my purpose, which was to add a warning that explains what just happened (why the installation failed), and would be an incremental improvement to the codebase.

@bcoca Thank you for helping me install Ansible from source code. Is there any documentation on how to update the changelog?   I looked at https://docs.ansible.com/ansible/latest/community/contributor_path.html and the dev guide.

Also, I see there is already an integration test for invalid filenames: https://github.com/ansible/ansible/blob/116948cd1468b8a3e34af8e3671f4089e1f7584c/test/integration/targets/ansible-galaxy-role/tasks/main.yml#L28-L31

Do you still expect me to add a test for my changeset?  I'm happy to learn how to do that if so.

**Comment by bcoca** [comment]
https://docs.ansible.com/ansible/latest/community/development_process.html#creating-changelog-fragments

**Comment by atsalolikhin-spokeo** [comment]
@bcoca I've added a changelog fragment; and I see we already have a test for invalid filenames: https://github.com/ansible/ansible/blob/116948cd1468b8a3e34af8e3671f4089e1f7584c/test/integration/targets/ansible-galaxy-role/tasks/main.yml#L28-L31

Do you still feel it's necessary to add tests to this PR?  (I'd be happy to do so if needed.)

Otherwise this changeset is ready for re-review, please.

Thanks very much for your help along the way!

**Comment by atsalolikhin-spokeo** [comment]
@bcoca Do you have any concerns about this changeset as it stands today?   :)  This is my first PR to the project, it's not my intention to make a major rewrite of any part (I saw Alexander's comment in commit 1c83672532330d0d26b63f8a97ee703e7fc697df that the entire module needs a rewrite), I just want to drop the user some breadcrumbs (a hint) about why role installation is failing when the role contains Java inner classes (files with `$` in the filenames).

**Review by bcoca — APPROVED** [review]
while we still need to add code to handle the exception, this is still an improvement and can be merged independantly

**Comment by atsalolikhin-spokeo** [comment]
Thank you very much @bcoca for orienting me to Ansible development and helping me with my first PR.  :)  Win!

**Comment by bcoca** [comment]
thank you for your contribution

**Pull request merged**